Add Vector 4

ejona86 opened the pull request:
This is the follow-up to #10697. See `vector4.cpp` for main TODO. ~Floppy support is cheating but quite functional in practice.~ (implemented) ~I sort of expect to add MFM HDD support before swapping to the real floppy infrastructure.~ (I failed to notice how IBM-centric the formatting is. I realized it would need most of the same stuff as floppy.)

~I am working to dump the ROMs from a Vector 4 that CP/M-86 and MS-DOS were used on; I'm hoping it is a newer ROM and that fixes the booting issue for those OSes.~ (Yep, old roms didn't support it. Booting gets further, but still not working.)

cuavas reviewed (commented):
Given we already have a floppy system and an S100 bus, I think it would be better if these facilities could be used upfront rather than adding bespoke implementations that will need to be rewritten anyway.

ejona86 reviewed (commented):
> Given we already have a floppy system and an S100 bus, I think it would be better if these facilities could be used upfront rather than adding bespoke implementations that will need to be rewritten anyway.

There's very little wasted work to ignore the existing floppy and S100 code. For the floppy, I read from a position from a file and need to keep track of the current track. I would have done that anyway because there's no way I hate myself enough to try and debug the floppy stuff at the same time as the rest of the system and the code was easy. The code is negligible compared to the learning and code needed for the new floppy system. Avoiding S-100 only took `map(0xc0, 0xc3).mirror(0xff00).rw("dualmode", FUNC(vector_dualmode_device::read), FUNC(vector_dualmode_device::write));`. I see those shortcuts as a huge win to have _anything_ working; mame was quite opaque for me and I didn't have much confidence unless software was running and working.

ejona86 commented on src/mame/vectorgraphic/sbcvideo.cpp:
```diff
@@ -0,0 +1,173 @@
+
+	screen_device &screen(SCREEN(config, "screen", SCREEN_TYPE_RASTER));
+	screen.set_refresh_hz(60);
+	screen.set_screen_update(crtc, FUNC(c6545_1_device::screen_update));
```
I saw that the Video details in the System Information changed on initial startup vs when running. I used that to configure `set_raw()` as `set_size()` didn't seem to influence it. `-nomax` still doesn't work for me, but `-window` does. I see no obvious difference in the window size based on my configuration of `screen` in the code, but I figure I'd need to configure how pixels are stretched or some such.

ejona86 commented:
@cuavas, I addressed all the concerns except the comment about the build system which I don't understand what you mean. I have MFM HDD support queued for after this goes in.

cuavas commented on src/devices/bus/s100/vectordualmode.h:
```diff
@@ -0,0 +1,55 @@
+	emu_timer *m_byte_timer;
+	emu_timer *m_sector_timer;
+	uint16_t m_pending_byte;
+	unsigned int m_pending_size;
```
Please use an explicitly sized integer for this member so it can be used portably in save states.

cuavas commented on src/lib/formats/vgi_dsk.cpp:
```diff
@@ -0,0 +1,144 @@
+
+*********************************************************************/
+
+#include <cstring>
+
```
Please put standard library headers after project headers.  But are you actually using `<cstring>` at all?  I didn’t see any uses of it in a quick read of the file.  I saw use of `<vector>` and `<cstdint>` which MAME relies on pretty heavily.

cuavas commented on src/lib/formats/vgi_dsk.cpp:
```diff
@@ -0,0 +1,144 @@
+			for (int sector = 0; sector < 16; sector++) {
+				for (int i = 0; i < 40; i++)
+					mfm_w(buf, 8, 0, half_bitcell_size);
+				std::size_t actual;
+				io.read(sector_bytes, std::size(sector_bytes), actual);
```
Please check for read errors and return `false` to avoid weird things happening if a read fails.

cuavas commented on src/lib/formats/vgi_dsk.cpp:
```diff
@@ -0,0 +1,144 @@
+	if (!fmt.head_count) {
+		int heads, tracks;
+		image->get_actual_geometry(tracks, heads);
+		printf("heads: %d tracks %d\n", heads, tracks);
```
This `printf` shouldn’t be here – maybe use `LOG_FORMATS` instead, and make the message give a bit more context?

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+
+****************************************************************************/
+
+#include "emu.h"
+
+#include "bus/s100/vectordualmode.h"
+#include "formats/vgi_dsk.h"
```
Please make it clearer that the unit header is local, i.e. format it like:
```c++
#include "emu.h"
#include "vectordualmode.h"

#include "formats/vgi_dsk.h"
```

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+	}
+}
+
+bool s100_vector_dualmode_device::hdd_selected()
+{
+	return m_drive == 0 && false;
+}
```
Please put a TODO comment on the expression here to indicate that it’s disabled pending adding hard disk support.

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+	// 7200-1200-02-1 page 16 (1-10)
+	uint8_t data;
+	if (offset == 0xc0) { // status (0) port
+		bool write_protect; // FDD
+		bool ready; // HDD
+		bool track0;
+		bool write_fault = false; // HDD
+		bool seek_complete; // HDD
+		bool loss_of_sync; // HDD
+		if (hdd_selected()) {
+			write_protect = false;
+			ready = true;
+			track0 = false;
+			seek_complete = true;
+			loss_of_sync = true;
+		} else {
+			floppy_image_device* flop = m_floppy[m_drive]->get_device();
+			write_protect = flop && flop->wpt_r();
+			ready = false;
+			track0 = flop && !flop->trk00_r();
+			seek_complete = false;
+			loss_of_sync = false;
+		}
+
+		data = write_protect | (ready << 1) | (track0 << 2)
+		    | (write_fault << 3) | (seek_complete << 4) | (loss_of_sync << 5)
+			| 0xc0;
```
Please don’t directly use mathematical operators on `bool` values.  Either turn the variables into `uint8_t` and assign them 0 or 1, or build up the result like `... (ready ? 0x02 : 0x00) | (track0 ? 0x04 : 0x00) ...` etc.  Some compilers and analysis tools produce warnings for using Boolean in integer context.

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+		    | (write_fault << 3) | (seek_complete << 4) | (loss_of_sync << 5)
+			| 0xc0;
+	} else if (offset == 0xc1) { // status (1) port
+		bool floppy_disk_selected;
+		bool controller_busy = m_sector_timer->enabled();
+		bool motor_on; // FDD
+		bool type_of_hard_disk = true;
+		if (hdd_selected()) {
+			floppy_disk_selected = false;
+			motor_on = false;
+		} else {
+			floppy_disk_selected = true;
+			motor_on = m_motor_on_timer->enabled();
+		}
+		data = floppy_disk_selected | (controller_busy << 1) | (motor_on << 2)
+		    | (type_of_hard_disk << 3) | 0xf0;
```
Same here, please rework this block to avoid using Boolean in integer context.

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+		}
+		data = floppy_disk_selected | (controller_busy << 1) | (motor_on << 2)
+		    | (type_of_hard_disk << 3) | 0xf0;
+	} else if (offset == 0xc2) { // data port
+		data = m_ram[m_cmar++];
+		m_cmar &= 0x1ff;
+	} else if (offset == 0xc3) { // reset port
+		m_cmar = 0;
+		data = 0xff;
+	} else {
```
You need to guard against debugger accesses on reads with side effects.  You need to check `if (!machine().side_effects_disabled())` before modifying contents of `m_cmar` here or bad things can happen when a debugger memory window is open.

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+
+TIMER_CALLBACK_MEMBER(s100_vector_dualmode_device::sector_cb)
+{
+	switch (static_cast<sector_timer_state>(param)) {
+	case SECTOR_START:
```
Just a nitpick, but since `sector_timer_state` is an unscoped enum, you don’t actually need the cast here at all, and even if you want to use it, a function syntax cast will work fine with less visual clutter (i.e. `switch (sector_timer_state(param)) {` is equivalent with less clutter).

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,315 @@
+	}
+}
+
+void s100_vector_dualmode_device::device_start()
+{
+	m_motor_on_timer = timer_alloc(FUNC(s100_vector_dualmode_device::motor_off), this);
+	m_byte_timer = timer_alloc(FUNC(s100_vector_dualmode_device::byte_cb), this);
+	m_sector_timer = timer_alloc(FUNC(s100_vector_dualmode_device::sector_cb), this);
+
+	save_item(NAME(m_ram));
+	save_item(NAME(m_cmar));
+	save_item(NAME(m_drive));
+	save_item(NAME(m_sector));
+	save_item(NAME(m_read));
+	save_item(NAME(m_pending_byte));
+	save_item(NAME(m_pending_size));
+}
```
Do you need to do something to register the PLL state for save states?

cuavas replied:
I guess leave it for now if nothing else is doing it properly.  You can implement `device_post_load` if you want to get a call after a saved state is restored.

cuavas reviewed (changes requested):
Sorry for the delay reviewing this, it’s been a very hectic couple of months in MAME development.  This looks like it’s going in the right direction now.  The issues I noticed are all fairly minor checklist type stuff.

If @galibert could look over the floppy format class I’d appreciate it.

cuavas commented on src/devices/bus/s100/vectordualmode.cpp:
```diff
@@ -0,0 +1,326 @@
+			m_cmar++;
+			m_cmar &= 0x1ff;
+		}
+	} else if (offset == 0xc3) { // reset port
+		m_cmar = 0;
+		data = 0xff;
+	} else {
```
You also need to check `machine().side_effects_disabled()` before resetting `m_cmar` in this case.

Tafoid commented:
This commit is missing the associated data (newly dumped romsets?) to support vector4.
In particular, we haven't been able to validate/find the following:

vector4: Unable to find matching ROM u35_sbc_v111_4f11.bin (fcfd42c6)
vector4: Unable to find matching ROM sbcexecf.bin (738e10b5)

Please send these to the address you sent the first batch of vector4 requirements in support of your other PR:  https://github.com/mamedev/mame/pull/10697

Thanks for your contributions!
